# Vectors

Congratulations on starting your very own landscaping company!

Thanks to R and your new data-analytical skills, you will learn how to uplift your profits and fire off your career like a true professional. This chapter will show how you can easily keep track of your profits and how you can do some simple analyses on past actions. Next stop, .... self made!!

#### Question 1

Do you still remember what you have learned in the first chapter? Assign the value `"Go!"` to the variable `Lets_get_started`. Remember: R is case sensitive!
- Note: Just type the following line: Lets_get_started <- "Go!"

In [39]:
#your code here


# Problem 1

In this problem we are going to work on naming vectors. Let us focus first!

On your way to riches, you will make extensive use of vectors. Vectors are one-dimension arrays that can hold numeric data, character data, or logical data. In other words, a vector is a simple tool to store data. For example, you can store your daily gains and losses in the casinos.

#### Instructions

In R, you create a vector with the combine function [`c()`](http://www.rdocumentation.org/packages/base/functions/c). You place the vector elements separated by a comma between the parentheses. For example:
- Note: you can use either the '=' or '<-' operator to assign values to a variable, for this lab we will be using the '<-' operator.

In [40]:
#Example

numeric_vector = c(1, 10, 49)
character_vector <- c("a", "b", "c")
print(numeric_vector)
print(character_vector)

[1]  1 10 49
[1] "a" "b" "c"


Once you have created these vectors in R, you can use them to do calculations.

#### Question 1

Complete the code such that `boolean_vector` contains the three elements: `TRUE`, `FALSE` and `TRUE` (in that order).
- Note: Assign c(TRUE, FALSE, TRUE) to the variable boolean_vector with the <- operator.

In [41]:
#your code here


# Problem 2

After six months of work and still zero Ferraris in your garage, you decide that it is time to start using your data analytical superpowers.

You don't know whether to focus on mulching or cutting grass, so you decide to track profit and loss tied to each seperately over the next six months, from March to August. 

For `mulch_vector`:

- March: Profit 85
- April: Profit 200
- May: Profit 225
- June: Loss 100
- July: Profit 40
- August: Loss 60

For `grasscut_vector`:

- March: Profit 150
- April: Loss 30
- May: Profit 170
- June: Loss 90
- July: Loss 50
- August: Profit 225

To be able to use this data in R, you decide to create the variables `mulch_vector` and `grasscut_vector`

#### Instructions

Assigned below is the profits/losses for mulching to the variable `mulch_vector`.

In [42]:
# Mulching Profit/Losses from March to August
mulch_vector <- c(85, 200, 225, -100, 40, -60)

#### Question 1

Assign the profits/losses for grass cutting to the variable `grasscut_vector`.

In [43]:
# Grass cutting Profit/Losses from March to August

#your code here

# Problem 3

Problem 3 deals with naming vectors. 

As a data analyst, it is important to have a clear view on the data that you are using. Understanding what each element refers to is therefore essential.

In the previous exercise, we created a vector profits/losses over the last 6 months. Each vector element refers to a month of the year but it is hard to tell which element belongs to which month. It would be nice if you could show that in the vector itself.

You can give a name to the elements of a vector with the `names()` function. Have a look at this example:

```
some_vector <- c("John Doe", "Landscaper")
names(some_vector) <- c("Name", "Profession")
```

This code first creates a vector `some_vector` and then gives the two elements a name. The first element is assigned the name `Name`, while the second element is labeled `Profession`. Printing the contents to the console yields following output:

```
          Name     Profession
    "John Doe"  "Landscaper"
```

#### Instructions

The code below names the elements in mulch_vector with the six months
- Note: Make sure to use the same vector with the months as names. Remember that R is case sensitive!

In [44]:
# Mulching Profit/Losses from March to August
mulch_vector <- c(85, 200, 225, -100, 40, -60)

# Assign months as names of the mulch_vector
names(mulch_vector) <- c("March", "April", "May", "June", "July", 'August')
print(mulch_vector)

 March  April    May   June   July August 
    85    200    225   -100     40    -60 


 #### Question 1

On your own produce code that names the elements in the grasscut_vector with the six months. Then print the variable grasscut_vector in order to verify you have named the vector correctly. 

In [45]:
#your code here


If you want to become a good statistician, you have to become lazy. (If you are already lazy, chances are high you are one of those exceptional, natural-born statistical talents.)

In the previous exercises you probably experienced that it is boring and frustrating to type and retype information such as the months of the year. However, when you look at it from a higher perspective, there is a more efficient way to do this, namely, to assign the months vector to a **variable**!

Just like you did with your mulching and grass cutting profits/losses, you can also create a variable that contains the months of interest. This way you can use and re-use it.

#### Instructions 

The code below creates a variable months_vector that contains the months of interest.

In [46]:
#The variable months_vector
months_vector <- c("March", "April", "May", "June", "July", 'August')

The months_vector is used to more efficiently name the mulch_vector.

In [47]:
names(mulch_vector) <- months_vector
print(mulch_vector)

 March  April    May   June   July August 
    85    200    225   -100     40    -60 


#### Question 2

On your own use the months_vector to give names to the grasscut_vector. Then print the variable grasscut_vector in order to verify you have named the vector correctly. 

In [48]:
#your code here


# Problem 4


Now that you have the mulching and grass cutting profits/losses as named vectors, you can start doing some data analytical magic.

You want to find out the following type of information:

- How much has been your overall profit or loss per month over the past six months?
- Have you lost money over the 6 months in total?
- Which is more profitable mulching or grass cutting?

To get the answers, you have to do arithmetic calculations on vectors.

#### Instructions

It is important to know that if you sum two vectors in R, it takes the element-wise sum. For example, the following three statements are completely equivalent.

In [49]:
c(1, 2, 3) + c(4, 5, 6)

[1] 5 7 9

In [50]:
c(1 + 4, 2 + 5, 3 + 6)

[1] 5 7 9

You can also do the calculations with variables that represent vectors.
- Note: Below the sum of vectors a and b were taken and assigned to the c vector. 

In [51]:
a <- c(1, 2, 3)
b <- c(4, 5, 6)
c <- a + b
c

[1] 5 7 9

#### Question 1

Take the sum of the variables A_vector and B_vector and assign it to total_vector. Inspect the result by printing out total_vector.

In [52]:
A_vector <- c(1, 2, 3)
B_vector <- c(4, 5, 6)

In [53]:
#your code here


Now you understand how R does arithmetic with vectors, it is time to get those Ferraris in your garage! First, you need to understand what the overall profit or loss per month was. The total six month profit/loss is the sum of the profit/loss you realized on mulching per month, and the profit/loss you realized on grass cutting per month.

In R, this is just the sum of `grasscut_vector` and `mulch_vector`.

#### Question 2

Assign to the variable `total_monthly` how much you profited or lost in total (mulching and grass cutting combined). Print out the total_monthly variable to view the monthly results.
- Note: Similar to the previous exercise, assign the sum of two vectors to a new variable, `total_monthly`.

In [54]:
#your solution here


#### Question 3

In which month were you the most profitable, which month did you take the biggest loss? 

In [55]:
#Month of Most Profit

#Month of Most lost


# Problem 5

Based on the previous analysis, it looks like you had a mix of good and bad months. But how did you over the 6 months in total and for each type of work?

A function that helps you to answer this question is [`sum()`](http://www.rdocumentation.org/packages/base/functions/sum). It calculates the sum of all elements of a vector.

#### Instructions 

The sum() function below calculates the total amount of money you have profited or lost in mulching over the 6 months and places it in a new variable called total_mulch.

In [56]:
total_mulch <- sum(mulch_vector)
total_mulch

[1] 390

Wow! Over the six months you profited $390 mulching congratulations!

#### Question 1

Calculate the total amount of money that you have profited and lost in grass cutting and assign to the variable total_grasscut. Then print out total_grasscut and comment on if you either profited or lost over the 6 months cutting grass and how much.

In [57]:
#your code here

#Comment... did you profit? If so, how much?


#### Question 2

Now that you have the totals for grass cutting and mulching, you can easily calculate total_6months (which is the sum of all the profits of the 6 months). Print out total_6months and comment on your results.

In [58]:
#your code here

#Comment... did you profit? If so, how much?


A quick way to analyze which method of landscaping is more profitable is to use a boolean statement that returns a logical value (`TRUE` or `FALSE`). 

#### Instructions

To check if 6 is larger than 5, you type `6 > 5`. 

In [59]:
6 > 5

[1] TRUE

As you can see, 6 is obviously greater than 5 so the statement reads TRUE. 

#### Question 3

Check if your total profits in mulching are higher than for grass cutting by using a comparison. Simply print out the result of this comparison. What do you conclude, should you focus on mulching or grass cutting?
- Note: The output of your answer should either read TRUE or FALSE. 

In [60]:
#your code here

#conclusion here


# Problem 6

It appears that mulching is more your cup of tea than cutting grass.

Another possible route for investigation is your performance at the first month compared to the performance of the last month. 

To answer that question, you only want to focus on a selection of the `total_vector`. In other words, our goal is to select specific elements of the vector. To select elements of a vector (and later matrices, data frames, ...), you can use square brackets. Between the square brackets, you indicate what elements to select. For example, to select the first element of the vector, you type `mulch_vector[1]`. To select the second element of the vector, you type `mulch_vector[2]`, etc. Notice that the first element in a vector has index 1, not 0 as in many other programming languages.

#### Instructions

The code below will select the profit/loss from August from the mulch vector. Saving it to a new variable mulch_August.

In [61]:
mulch_August <- mulch_vector[6]
print(mulch_August)

August 
   -60 


#### Question 1

Assign the mulching results of May to the variable mulch_May. Print mulch_May.
- Note: May is the third element of mulch_vector.

In [62]:
#your code here


#### Instructions 

To select multiple elements from a vector, you can add square brackets at the end of it. You can indicate between the brackets what elements should be selected. For example: suppose you want to select the first and the fifth day of the week: use the vector `c(1, 5)` between the square brackets. For example, the code below selects the first and fifth element of `mulch_vector`.

In [63]:
print(mulch_vector[c(1, 5)])

March  July 
   85    40 


How about analyzing your summer mulching results?

#### Question 2

Assign the mulching results of June, July, and August to the variable `mulch_summer`.

In [64]:
#your code here


#### Question 3

Use the sum() function to figure out how much you profited or lost mulching during the summer months. 

In [65]:
#your code here

#Did you profit in the summer months mulching? If so, how much? If not, how much did you lose?


Selecting multiple elements of `mulch_vector` with `c(4, 5, 6)` is not very convenient. Many statisticians are lazy people by nature, so they created an easier way to do this: `c(4, 5, 6)` can be abbreviated to`4:6`, which generates a vector with all natural numbers from 4 up to 6.

#### Instructions 

So, another way to find the summer results is `mulch_vector[4:6]`. Notice how the vector `4:6` is placed between the square brackets to select element 4 up to 6.

In [66]:
mulch_summer <- mulch_vector[4:6]
print(mulch_summer)

  June   July August 
  -100     40    -60 


#### Question 4

Assign to `grasscut_selection_vector` the grass cutting results from April up to July; make use of `:` if it makes things easier for you. Print `grasscut_selection_vector` to check and make sure the new vector contains all the appropriate months. 

In [67]:
#your code here


Another way to tackle the previous exercise is by using the names of the vector elements (March, April, ...) instead of their numeric positions. For example,

```
mulch_vector["March"]
```

will select the first element of `mulch_vector` since `"March"` is the name of that first element.

#### Instructions

Just like you did in the previous exercise with numerics, you can also use the element names to select multiple elements, for example. The code below selects the months March and April from the mulch_vector.

In [68]:
print(mulch_vector[c("March","April")])

March April 
   85   200 


To find the average profit in these first two months use the [`mean()`](http://www.rdocumentation.org/packages/base/functions/mean) function as demonstrated below.

In [69]:
mean(mulch_vector[c("March","April")])

[1] 142.5

The average profit for mulching in the first two months is $142.50. Increase sample size and add an extra month and to see if you get a different result. 

#### Question 5

Select the first three elements in mulch_vector by using their names. Assign the result of the selection to `mulch_start`. Then use the mean() function to analyze the mulch_start vector.

In [70]:
#your code here


In [71]:
#your code here

#mean()

# Problem 7


By making use of comparison operators, we can approach the previous question in a more proactive way.

The (logical) comparison operators known to R are:

- `<` for less than
- `>` for greater than
- `<=` for less than or equal to
- `>=` for greater than or equal to
- `==` for equal to each other
- `!=` not equal to each other

#### Instructions

As seen in the previous chapter, stating 6 > 5 returns TRUE. The nice thing about R is that you can use these comparison operators also on vectors.

In [72]:
c(4, 5, 6) > 5

[1] FALSE FALSE  TRUE

This command tests for every element of the vector if the condition stated by the comparison operator is TRUE or FALSE.

#### Question 1

Check which elements in mulch_vector are positive (i.e. > 0) and assign this to selection_vector. Print out selection_vector so you can inspect it. The printout tells you whether you profited (TRUE) or lost (FALSE) any money for each month.
- Note:  In order to check for which months your mulching gains are positive, R should check for each element of mulch_vector whether it is larger than zero. some_vector > 0 is the way to tell R what you are after.

In [73]:
#your code here


Working with comparisons will make your data analytical life easier. Instead of selecting a subset of months to investigate yourself (like before), you can simply ask R to return only the months where you realized a positive return for mulching.

In the previous exercises you used `selection_vector <- mulch_vector > 0` to find the days on which you had a positive mulching  return. Now, you would like to know not only the months you profited, but also how much you profited on those months.

#### Instructions

You can select the desired elements, by putting `selection_vector` between the square brackets that follow `mulch_vector`.
- Note: All values in the mulch_vector where 'mulch_vector > 0' is TRUE will be printed.

In [74]:
print(mulch_vector[mulch_vector > 0])

March April   May  July 
   85   200   225    40 


Notice how because mulch_vector > 0 is assigned to 'selection_vector' we get the same results. 

In [77]:
selection_vector <- mulch_vector > 0
print(mulch_vector[selection_vector])

March April   May  July 
   85   200   225    40 


#### Question 2

Use `selection_vector` in square brackets to assign the amounts that you profited on the profitable months to the variable `mulch_profit_months`.

In [ ]:
#your code here


Just like you did for mulching, you also want to know those months where you realized a positive return for grass cutting.

#### Question 3

Create the variable `selection_vector`, this time to see if you made profit with grass cutting for different months.

In [ ]:
#your code here


#### Question 4

Assign the amounts that you made on the months that you ended positively for grass cutting to the variable `grasscut_profit_months`. This vector thus contains the profitable months of `grasscut_vector`.


In [ ]:
#your code here


<hr>
#### For an .ipynb file of this lab follow <a href="https://github.com/anthonyiarussi/Introduction-to-Data-Science/tree/master/Student%20Labs/Labs%20.ipynb/Lab2" target="_blank">this link</a>.

#### For a .Rmd file of this lab follow <a href="https://github.com/anthonyiarussi/Introduction-to-Data-Science" target="_blank">this link</a>.
<hr>
#### Go to the <a href="https://github.com/anthonyiarussi/Introduction-to-Data-Science" target="_blank">GitHub repository</a>, where you can download a copy of the lab for yourself!